invokeai/app/services/download/download_base.py

```diff
@@ -0,0 +1,200 @@
+    job_started: Optional[str] = Field(default=None, description="Timestamp for when the download job started")
+    job_ended: Optional[str] = Field(
+        default=None, description="Timestamp for when the download job ende1d (completed or errored)"
+    )
+    bytes: int = Field(default=0, description="Bytes downloaded so far")
+    total_bytes: int = Field(default=0, description="Total file size (bytes)")
+
+    # set when an error occurs
+    error_type: Optional[str] = Field(default=None, description="Name of exception that caused an error")
+    error: Optional[str] = Field(default=None, description="Traceback of the exception that caused an error")
+
+    # internal flag
+    _cancelled: bool = PrivateAttr(default=False)
+
+    # optional event handlers passed in on creation
+    _on_start: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+    _on_progress: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+    _on_complete: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+    _on_error: Optional[DownloadEventHandler] = PrivateAttr(default=None)
+
+    def __le__(self, other: "DownloadJob") -> bool:
+        """Return True if this job's priority is less than another's."""
+        return self.priority <= other.priority
+
+    def cancel(self) -> None:
+        """Call to cancel the job."""
+        self._cancelled = True
+
+    # cancelled and the callbacks are private attributes in order to prevent
+    # them from being serialized and/or used in the Json Schema
+    @property
+    def cancelled(self) -> bool:
+        """Call to cancel the job."""
+        return self._cancelled
+
+    @property
+    def on_start(self) -> Optional[DownloadEventHandler]:
+        """Return the on_start event handler."""
+        return self._on_start
+
+    @property
+    def on_progress(self) -> Optional[DownloadEventHandler]:
+        """Return the on_progress event handler."""
+        return self._on_progress
+
+    @property
+    def on_complete(self) -> Optional[DownloadEventHandler]:
+        """Return the on_complete event handler."""
+        return self._on_complete
+
+    @property
+    def on_error(self) -> Optional[DownloadEventHandler]:
+        """Return the on_error event handler."""
+        return self._on_error
+
+    def set_callbacks(
+        self,
+        on_start: Optional[DownloadEventHandler] = None,
+        on_progress: Optional[DownloadEventHandler] = None,
+        on_complete: Optional[DownloadEventHandler] = None,
+        on_error: Optional[DownloadEventHandler] = None,
+    ) -> None:
+        """Set the callbacks for download events."""
+        self._on_start = on_start
+        self._on_progress = on_progress
+        self._on_complete = on_complete
+        self._on_error = on_error
+
+
+class DownloadQueueServiceBase(ABC):
+    """Multithreaded queue for downloading models via URL."""
+
+    @abstractmethod
+    def download(
```

Comment:
I wonder if the destination should have something like `.progress` appended to the filename while downloading, to indicate that it is an incomplete file.

Reply:
Done.